File: ios/SignalKit/Sources/SignalKit/Sync/SyncEngine.swift
import Foundation
import Network

/// Drives the offline-first upload loop: claims a batch of pending rows,
/// uploads it, and reconciles the queue based on the server's per-record
/// accept/reject response. Runs on its own serial queue so capture threads
/// never block on network I/O.
public final class SyncEngine {
    private let store: SignalStore
    private let apiClient: APIClient
    private var deviceID: UUID
    private let appVersion: String
    private let osVersion: String

    private let syncQueue = DispatchQueue(label: "com.signals.syncEngine")
    private let pathMonitor = NWPathMonitor()
    private var isSyncing = false
    private var backoff = Backoff()

    /// Fired after each sync pass (success, failure, or no-op) so observers
    /// can refresh UI state that depends on the queue -- `triggerSync()`
    /// only dispatches the pass asynchronously and returns immediately, so
    /// polling queue state right after calling it reflects the state from
    /// *before* the pass ran, not after.
    public var onPassCompleted: (() -> Void)?

    /// Plan calls for retaining uploaded rows ~24-48h before pruning, purely
    /// to bound local DB growth independent of upload success rate.
    private let uploadedRetention: TimeInterval = 48 * 3600

    public init(store: SignalStore, apiClient: APIClient, deviceID: UUID, appVersion: String, osVersion: String) {
        self.store = store
        self.apiClient = apiClient
        self.deviceID = deviceID
        self.appVersion = appVersion
        self.osVersion = osVersion
    }

    public func start() {
        pathMonitor.pathUpdateHandler = { [weak self] path in
            if path.status == .satisfied {
                self?.triggerSync()
            }
        }
        pathMonitor.start(queue: syncQueue)
    }

    public func stop() {
        pathMonitor.cancel()
    }

    /// Called on app foreground/background-transition, a `BGProcessingTask`
    /// firing, or a manual/debug trigger -- all funnel into the same
    /// serialized sync loop.
    public func triggerSync() {
        syncQueue.async { [weak self] in
            self?.runSyncLoop()
        }
    }

    /// Adopts a device ID from QR/deep-link provisioning. Applied on the sync
    /// queue so it's consistent with `runSyncLoop`'s reads; queued rows then
    /// upload under the new identity (per-row payloads carry no device_id).
    public func updateDeviceID(_ id: UUID) {
        syncQueue.async { [weak self] in
            self?.deviceID = id
        }
    }

    private func runSyncLoop() {
        guard !isSyncing else { return }
        isSyncing = true
        defer {
            isSyncing = false
            pruneOldUploaded()
            onPassCompleted?()
        }

        while let pendingCount = try? store.pendingCount(), pendingCount > 0 {
            let batchID = UUID().uuidString
            guard let rows = try? store.claimBatch(batchID: batchID), !rows.isEmpty else { break }

            let payload = UploadBatchBuilder.build(
                deviceID: deviceID,
                appVersion: appVersion,
                osVersion: osVersion,
                batchID: batchID,
                rows: rows.map {
                    (id: $0.id, capturedAt: $0.capturedAt, signalType: $0.signalType, payloadJSON: $0.payloadJSON,
                     lat: $0.lat, lon: $0.lon, horizontalAccuracy: $0.horizontalAccuracy, locationAgeSeconds: $0.locationAgeSeconds)
                }
            )

            let semaphore = DispatchSemaphore(value: 0)
            var uploadResult: Result<BatchUploadResponse, Error>?
            Task {
                do {
                    uploadResult = .success(try await apiClient.uploadBatch(payload))
                } catch {
                    uploadResult = .failure(error)
                }
                semaphore.signal()
            }
            semaphore.wait()

            switch uploadResult {
            case .success(let response):
                backoff.reset()
                try? store.markUploaded(ids: response.accepted)
                let permanentlyRejected = response.rejected.map(\.id)
                try? store.markFailed(ids: permanentlyRejected, permanent: true)

            case .failure(let error):
                try? store.markFailed(ids: rows.map(\.id), permanent: false)
                if case APIClientError.unauthorized = error {
                    // Re-auth is out of scope for the sync loop -- surface via
                    // a delegate/notification in a future iteration; for now
                    // stop this pass so we don't spin against a bad token.
                    return
                }
                // Schedule the retry instead of blocking this thread on
                // Thread.sleep: blocking would hold `isSyncing` for up to the
                // 10-minute backoff cap, silently swallowing any foreground
                // or manual "sync now" trigger that arrives in the meantime.
                scheduleRetry(after: backoff.next())
                return

            case .none:
                return
            }
        }
    }

    private func scheduleRetry(after delay: TimeInterval) {
        syncQueue.asyncAfter(deadline: .now() + delay) { [weak self] in
            self?.triggerSync()
        }
    }

    private func pruneOldUploaded() {
        try? store.pruneUploaded(olderThan: Date().addingTimeInterval(-uploadedRetention))
    }
}

/// Exponential backoff with jitter, base 5s capped at 10min, reset on success.
private struct Backoff {
    private var attempt = 0
    private let base: TimeInterval = 5
    private let cap: TimeInterval = 600

    mutating func next() -> TimeInterval {
        defer { attempt += 1 }
        let exponential = min(cap, base * pow(2, Double(attempt)))
        return exponential * Double.random(in: 0.8...1.2)
    }

    mutating func reset() {
        attempt = 0
    }
}

File: android/app/src/main/java/com/naberconsulting/signals/sync/SyncEngine.kt
package com.naberconsulting.signals.sync

import android.content.Context
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkRequest
import android.util.Log
import com.naberconsulting.signals.queue.SignalStore
import java.util.UUID
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import kotlin.math.min
import kotlin.math.pow
import kotlin.random.Random

/**
 * Drives the offline-first upload loop: claims a batch of pending rows,
 * uploads it, and reconciles the queue from the server's per-record
 * accept/reject response. Runs on a single-thread executor so capture threads
 * never block on network I/O -- the Android counterpart to the iOS SyncEngine.
 */
class SyncEngine(
    context: Context,
    private val store: SignalStore,
    private val apiClient: ApiClient,
    private val meta: MetaProvider,
) {
    /** Batch metadata resolved lazily so device_id/app/os stay current. */
    interface MetaProvider {
        fun deviceId(): String
        fun appVersion(): String
        fun osVersion(): String
    }

    /** Fired after each pass (success/failure/no-op) so UI can refresh queue state. */
    var onPassCompleted: (() -> Unit)? = null

    private val appContext = context.applicationContext
    private val executor = Executors.newSingleThreadScheduledExecutor()
    private val connectivity = appContext.getSystemService(ConnectivityManager::class.java)
    @Volatile private var isSyncing = false
    private var attempt = 0

    private val networkCallback = object : ConnectivityManager.NetworkCallback() {
        override fun onAvailable(network: Network) = triggerSync()
    }

    fun start() {
        val request = NetworkRequest.Builder()
            .addCapability(android.net.NetworkCapabilities.NET_CAPABILITY_INTERNET)
            .build()
        connectivity?.registerNetworkCallback(request, networkCallback)
    }

    fun stop() {
        runCatching { connectivity?.unregisterNetworkCallback(networkCallback) }
    }

    /** All triggers -- connectivity, app foreground, manual "sync now" -- funnel here. */
    fun triggerSync() {
        executor.execute { runSyncLoop() }
    }

    private fun runSyncLoop() {
        if (isSyncing) return
        isSyncing = true
        try {
            while (store.pendingCount() > 0) {
                val batchId = UUID.randomUUID().toString()
                val rows = store.claimBatch(batchId)
                if (rows.isEmpty()) break

                val body = UploadBatchBuilder.build(
                    UploadBatchBuilder.Meta(meta.deviceId(), meta.appVersion(), meta.osVersion(), batchId),
                    rows,
                )
                try {
                    val response = apiClient.uploadBatch(body)
                    attempt = 0
                    store.markUploaded(response.accepted)
                    store.markFailed(response.rejectedIds, permanent = true)
                } catch (e: UnauthorizedException) {
                    // Re-auth is out of scope for the loop; stop rather than
                    // spin against a bad token. Rows revert to pending.
                    store.markFailed(rows.map { it.id }, permanent = false)
                    Log.w(TAG, "upload unauthorized; pausing sync pass")
                    return
                } catch (e: Exception) {
                    store.markFailed(rows.map { it.id }, permanent = false)
                    scheduleRetry(nextBackoffSeconds())
                    Log.w(TAG, "upload failed: ${e.message}; retry scheduled")
                    return
                }
            }
        } finally {
            isSyncing = false
            pruneOldUploaded()
            onPassCompleted?.invoke()
        }
    }

    private fun scheduleRetry(delaySeconds: Double) {
        executor.schedule({ runSyncLoop() }, (delaySeconds * 1000).toLong(), TimeUnit.MILLISECONDS)
    }

    /** Exponential backoff with jitter, base 5s capped at 10min. */
    private fun nextBackoffSeconds(): Double {
        val exponential = min(BACKOFF_CAP, BACKOFF_BASE * 2.0.pow(attempt))
        attempt += 1
        return exponential * Random.nextDouble(0.8, 1.2)
    }

    private fun pruneOldUploaded() {
        store.pruneUploaded(System.currentTimeMillis() - UPLOADED_RETENTION_MS)
    }

    companion object {
        private const val TAG = "SyncEngine"
        private const val BACKOFF_BASE = 5.0
        private const val BACKOFF_CAP = 600.0
        private const val UPLOADED_RETENTION_MS = 48L * 3600 * 1000
    }
}
